src/compile/legend/parse.ts

```diff
@@ -51,6 +54,10 @@ export function parseLegendForChannel(model: UnitModel, channel: NonPositionScal
   const fieldDef = model.fieldDef(channel);
   const legend = model.legend(channel);
 
+  if (!model.getScaleComponent(channel)) {
```

Comment:
Move the logic to `parseUnitLegend` so that the structure of both `parseUnitLegend` and `parseUnitAxis` are similar?